s30291-pj-StanislasHinsinger/zajecia5/zadanie6.c

```diff
@@ -0,0 +1,77 @@
+#include <stdio.h>
+#include <sys/wait.h>
+#include <sys/types.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+/*
+    DO OSOBY SPRAWDZAJACEJ:
+
+    Pomysl:
+    - dwa procesy dzici (sync_loop i input_loop)
+    - sync_loop = petla, ktora bedzie otwierac plik, czytac jego zawartosc,
+      sprawdzac ilosc lini czy jest wieksza od ostatniej zczytanej ilosci,
+      jezeli tak, to wypisac wszystkie linie przekraczajace ostatni zakres,
+      wyswietlic na ekranie, zamknac plik
+    - input_loop = petla, ktora bedzie w nieskonczonosc czekac na input
+      uzytkownika gdzie za kazdym razem jest prefix "nazwa_klienta: ",
+      gdy go dostanie to otwiera plik ("a+"), dodaje wprowadzony tekst,
+      zapisuje go, dokonuje ++ na ostatnie zczytanej ilosci lini.
+
+    Problemy:
+    - gdy dokona sie printf zostanie za kazdy razem dodane w linijce przed tekst
+      ktory mial byc z prefixu od sync_loop "nazwa_klienta: ", jezeli nawet
+      nie wprowadzac tego prefixu to wtedy i tak wyswietli text z wprowadzonego buffora.
+    - petla sync_loop i petla input_loop moga sie ze soba klocic jezeli chodzi o dostep do pliku
+      (trzeba delay zrobic, (usleep) nie wykrywalo mi, a sleep daje tylko sekundy co nie
+      rozwiazuje jakkolwiek sensownie problemu)
+    - mozna by bylo zrobic jakas glebsza logike powiazana z dzialaniem socketow (nie dokonca),
+      tylko ze na pliku, ze clienci dokonuja ciaglego "streamingu" tego co sie dzieje, taki
+      jakby live-update.
+
+    Zadanie jest nie wykonane, ze wzgledu na brak wystraczajacej wiedzy. (wielokrotnie scrapowalem ten plik)
+    Prosilbym o nakierowanie jakbym mogl te problemy ewentualnie rozwiazac, albo jaka inna droge moge obrac?
```

Comment:
Ważne, że przysiadłeś do tego sam i próbowałeś rozwiązać. Na pewno nauczyłeś się sporo.
A teraz kwestia rozwiązania:
Poczytaj o semaforach i plikach lock. Jeśli plik lock istnieje to wtedy nie wolno innemu procesowi niż ten który taki plik utworzył pisac do pliku. Gdy zapis jest kończony to plik lock powinien być usunięty. Wtedy wolno czytać do woli. Są dedykowane funkcje do tworzenia takich plików lock.

https://stackoverflow.com/a/33939497